app/src/ui/stashing/stash-diff-header.tsx

```diff
@@ -0,0 +1,99 @@
+import * as React from 'react'
+import { IStashEntry } from '../../models/stash-entry'
+import { Dispatcher } from '../dispatcher'
+import { Repository } from '../../models/repository'
+import { Button } from '../lib/button'
+import { ButtonGroup } from '../lib/button-group'
+import { PopupType } from '../../models/popup'
+import { Octicon, OcticonSymbol } from '../octicons'
+
+interface IStashDiffHeaderProps {
+  readonly stashEntry: IStashEntry
+  readonly repository: Repository
+  readonly dispatcher: Dispatcher
+  readonly isWorkingTreeClean: boolean
+}
+
+interface IStashDiffHeaderState {
+  readonly isRestoring: boolean
+}
+
+/**
+ * Component to provide the actions that can be performed
+ * on a stash while viewing a stash diff
+ */
+export class StashDiffHeader extends React.Component<
+  IStashDiffHeaderProps,
+  IStashDiffHeaderState
+> {
+  public constructor(props: IStashDiffHeaderProps) {
+    super(props)
+
+    this.state = {
+      isRestoring: false,
+    }
+  }
+
+  public render() {
+    const { isWorkingTreeClean } = this.props
+    const restoreMessage = isWorkingTreeClean ? (
+      <span className="text">
+        <strong>Restore</strong> will move your stashed files to the Changes
+        list.
+      </span>
+    ) : (
+      <>
+        <Octicon symbol={OcticonSymbol.alert} />
+        <span className="text">
+          Unable to restore stash when changes are present on your branch.
+        </span>
+      </>
+    )
+
+    // we pass `false` to `ButtonGroup` below because it assumes
+    // the "submit" button performs the destructive action.
+    // In this case the destructive action is performed by the
+    // non-submit button so we _lie_ to the props to get
+    // the correct button ordering
+    return (
+      <div className="header">
+        <h3>Stashed changes</h3>
+        <div className="row">
+          <ButtonGroup destructive={false}>
+            <Button
+              disabled={!isWorkingTreeClean || this.state.isRestoring}
+              onClick={this.onRestoreClick}
+              type="submit"
+            >
+              Restore
+            </Button>
+            <Button
+              disabled={this.state.isRestoring}
+              onClick={this.onDiscardClick}
+            >
+              Discard
+            </Button>
+          </ButtonGroup>
+          <div className="explanatory-text">{restoreMessage}</div>
+        </div>
+      </div>
+    )
+  }
+
+  private onDiscardClick = () => {
+    const { dispatcher, repository, stashEntry } = this.props
+    dispatcher.showPopup({
+      type: PopupType.ConfirmDiscardStash,
+      stash: stashEntry,
+      repository,
+    })
+  }
+
+  private onRestoreClick = async () => {
+    const { dispatcher, repository, stashEntry } = this.props
+
+    this.setState({ isRestoring: true }, async () => {
```

Comment:
I've not been able to trip this issue myself on macOS or Windows using build 692d7d9f87131d2114af5be78751abe5d46655c1 but my gut feeling is that somehow `onRestoreClick` can fire multiple times before `setState` executes?

Reply:
I have not been able to get this to occur again either and I had the same thought about onRestoreClick being fired multiple times before `setState` actually sets the state. 

Do you think the potential for multiple calls is because of React batching state updates?

Reply:
> Do you think the potential for multiple calls is because of React batching state updates?

The updates should be batched, but because we're hooking into the callback that code should run _after_ the state update has been applied. Maybe there's something about how `callback` is invoked (it expects a plain function, but we're using async/await) that's part of the problem?

I'd love to hear from @tierninho about whether there's some repro setup that we've overlooked...

Reply:
I don't technically need to await that function since I'm not doing any state updates after the fact. I'll push that change up in a bit, but 👍 on figuring out if there's something I'm missing that @tierninho is doing.